ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,499 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+        }
     }
 
+    /* We're done, release the lock */
     KeReleaseSpinLock(&PspQuotaLock, OldIrql);
     return STATUS_SUCCESS;
 }
 
-/*
- * Private helper to remove quota charge from the specified process quota.
- * Notes: Conceptually translation unit local/private.
+/**
+ * @brief
+ * Internal kernel function that provides the
+ * bulk logic of process quota returning. It
+ * returns (takes away) quotas back from a
+ * process and/or quota block, which is
+ * the opposite of charging quotas.
+ *
+ * @param[in] Process
+ * A process, represented as a EPROCESS object.
+ * This parameter is used to return the own
+ * process' quota usage.
+ *
+ * @param[in] QuotaBlock
+ * The quota block which quotas are to be returned.
+ * This block can either come from the process itself
+ * or from an object with specified quota charges.
+ *
+ * @param[in] QuotaType
+ * The quota type which quota in question is to
+ * be returned. The permitted types are PsPagedPool,
+ * PsNonPagedPool and PsPageFile.
+ *
+ * @param[in] Amount
+ * The amount of quota to be returned.
+ *
+ * @return
+ * Nothing.
  */
 VOID
 NTAPI
-PspReturnProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN PS_QUOTA_TYPE QuotaType,
-                                   IN SIZE_T    Amount)
+PspReturnProcessQuotaSpecifiedPool(
+    _In_opt_ PEPROCESS Process,
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock,
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _In_ SIZE_T Amount)
 {
     KIRQL OldIrql;
-    ASSERT(Process);
-    ASSERT(Process != PsInitialSystemProcess);
+    SIZE_T ReturnThreshold;
+    SIZE_T AmountToReturn = 0;
+    SIZE_T UpdatedQuotaUsage = 0;
+
+    /* Sanity checks */
     ASSERT(QuotaType < PsQuotaTypes);
     ASSERT(!(Amount & 0x80000000)); /* we need to be able to negate it */
 
+    /* If the caller gives no amount then something's wrong */
+    ASSERT(Amount != 0);
+
     /* Guard our quota in a spin lock */
     KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
 
-    if (Process->QuotaUsage[QuotaType] < Amount)
+    /* Does the caller return more quota than it was previously charged? */
+    if ((Process && Process->QuotaUsage[QuotaType] < Amount) ||
+        QuotaBlock->QuotaEntry[QuotaType].Usage < Amount)
     {
-        DPRINT1("WARNING: Process->QuotaUsage sanity check failed.\n");
+        /* It does, crash the system! */
+        KeBugCheckEx(QUOTA_UNDERFLOW,
+                     (ULONG_PTR)Process,
+                     (ULONG_PTR)QuotaType,
+                     Process ? (ULONG_PTR)Process->QuotaUsage[QuotaType] :
+                               QuotaBlock->QuotaEntry[QuotaType].Usage,
+                     (ULONG_PTR)Amount);
     }
-    else
+
+    /*
+     * We need to trim the quota limits based on the
+     * amount we're going to return quotas back.
+     */
+    if (QuotaType != PsPageFile &&
+        QuotaBlock != &PspDefaultQuotaBlock)
```

Comment:
Can we really get here with `PspDefaultQuotaBlock`? That block never actually gets anything charged against it IIUC, so I would imagine it would hit the bugcheck above, since `Usage` will always be zero

Reply:
I'm not sure what's exactly your point of this comment. This branch is about Return and Limit fields management of a quota block as one mustn't touch the default block's limits (otherwise it would not make sense logically to call `PspDefaultQuotaBlock` a default block with default limits if you touch them at this point). `Usage` field is handled outside of this branch.

Reply:
I'm suggesting that this point in the code may be unreachable if `QuotaBlock == &PspDefaultQuotaBlock`, because either
* this function is never called with that quota block, or
* if it was called this way, it would likely have bugchecked above.

I may be making a wrong assumption somewhere, but that's my reading of the code.